lib/web_catalog/api_extractor.es6.js

```diff
@@ -0,0 +1,279 @@
+  @param {String} name - the property name from its parent node. (not necessarily interface name)
+  @param {Array} visited - a list of visted properties
+  @param {JSON} opt - know if this object is firstlevel or a __proto__ property.
+*/
+function extractAPI(og, map, id, name, visited, opt) {
+  opt = opt || {};
+  var firstLevel = opt.firstLevel || false;
+  var proto = opt.proto || false;
+
+  // Return if this object is already visited, except for first level objects.
+  // Still visit it if this is a firstLevel object, since
+  // two different firstlevel object can reference to one object.
+  // We don't want to miss any of them.
+  if (visited.indexOf(id) !== -1 && !firstLevel) return;
+  // An object/function's Id must be greater than root's Id.
+  // And we don't want to revisit window object.
+  if (id <= rootId) return;
+  visited.push(id);
+  if (og.isFunction(id) && (proto || firstLevel)) {
+    // This object is a function and it is either a __proto__ or on first level.
+    // Function which is not a __protot__ or first level is not considered as an interface
+    // eg: chrome.webstore, chrome.runtime.
+    if (id === functionId) return; // Skip if function is Function.prototype;
+    let prototypeId = og.lookup('prototype', id);
+    let ownProperties = getObjectProperties(og, id);
+    let prototypeProps = [];
+    // If function exists as an first level interface, use its property name
+    // under window object as interface name, otherwise, use function name.
+    let interfaceName = firstLevel ? name : og.getFunctionName(id);
+    if (prototypeId) prototypeProps = getObjectProperties(og, prototypeId);
+    arrayMerge(ownProperties, prototypeProps);
+    let meaningfulProps = setMinus(ownProperties, blacklistFunctionAttr);
+    if (meaningfulProps.length > 0) {
+      // If this function and its prototype contains meaning ful properties.
+      if (!map[interfaceName]) map[interfaceName] = [];
+      arrayMerge(map[interfaceName], getObjectProperties(og, id),
+          getObjectProperties(og, prototypeId));
+    }
+    if (og.getPrototype(id)) {
+      extractAPI(og, map, og.getPrototype(id), `${name}__proto__`, visited, {proto: true});
+    }
+  } else if (!og.isFunction(id)) {
+    // Object is not a function.
+    let ownProperty = getObjectProperties(og, id);
+    if (setMinus(ownProperty, blacklistObjectAttr).length > 0) {
+      let className = getClassName(id, og);
+      if ((!className || className === "Object") && firstLevel) {
+        // If object is at first level and its class name is "Object",
+        // this object is considered as an object library.
+        let prototypeId = og.lookup('prototype', id);
+        // in Firefox Edge ..., objects can have prototype properties.
+        let prototypeProps = [];
+        if (prototypeId) {
+          prototypeProps = getObjectProperties(og, prototypeId);
+        }
+        // if (!map[name]) map[name] = [INTERFACE];
```

Comment:
Just commented-out from previous implementation? If so, we should just delete this line. If it's still here because sometimes you uncomment it for debugging, please add a comment above explaining how it's used for debugging.